libraries/AP_GPS/AP_GPS_UBLOX.cpp

```diff
@@ -580,11 +578,15 @@ AP_GPS_UBLOX::read(void)
         }
     }
 
-    numc = port->available();
-    for (int16_t i = 0; i < numc; i++) {        // Process bytes received
+    const uint16_t numc = MIN(port->available(), 8192U);
+    for (uint16_t i = 0; i < numc; i++) {        // Process bytes received
 
         // read the next byte
-        data = port->read();
+        const int16_t rdata = port->read();
+        if (rdata < 0) {
+            break;
```

Comment:
Isn't this a significant internal error if you are checking for it? We prechecked the available bytes, and no one else should be reading from our port. If we care enough to check for it we should flag it, and if it's not important enough to throw an error, why are we checking?

Reply:
We do this check all over the place in other serial drivers.  It would be a significant error in the code if it happened.

Reply:
internal error costs flash, this type of protection against too much cpu usage per loop is good practice, but doing an internal error on it is overkill
what it protects against is a bad UART peripheral (eg. massive voltage spike, uart reports huge amount of spuious data in DMA).